src/java.base/share/classes/sun/security/rsa/RSASignature.java

```diff
@@ -205,24 +208,29 @@ protected boolean engineVerify(byte[] sigBytes) throws SignatureException {
         }
         try {
             if (sigBytes.length != RSACore.getByteLength(publicKey)) {
-                throw new SignatureException("Signature length not correct: got " +
+                throw new SignatureException("Bad signature length: got " +
                     sigBytes.length + " but was expecting " +
                     RSACore.getByteLength(publicKey));
             }
+
+            // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
+            // Step 4 suggests comparing the encoded message
             byte[] digest = getDigestValue();
+            byte[] encoded = RSAUtil.encodeSignature(digestOID, digest);
+            byte[] padded = padding.pad(encoded);
             byte[] decrypted = RSACore.rsa(sigBytes, publicKey);
-            byte[] unpadded = padding.unpad(decrypted);
-            // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
-            // Step 4 suggests comparing the encoded message instead of the
-            // decoded, but some vendors might omit the NULL params in
-            // digest algorithm identifier.
-            byte[] decodedDigest = RSAUtil.decodeSignature(digestOID, unpadded);
-            return MessageDigest.isEqual(digest, decodedDigest);
+
+            boolean status = MessageDigest.isEqual(padded, decrypted);
+            if (!status) {
+                // fall back to the decode approach for max compatibility
+                byte[] unpadded = padding.unpad(decrypted);
+                if (unpadded != null) {
+                    status = MessageDigest.isEqual(digest,
+                            RSAUtil.decodeSignature(digestOID, unpadded));
+                }
+            }
+            return status;
         } catch (javax.crypto.BadPaddingException e) {
```

Comment:
Can this exception still be thrown?

Reply:
Yes, BadPaddingException is still thrown by RSACore class.

Reply:
That exception is thrown by RSACore.parseMsg() if the Message is larger than the modulus. I think it is at least debatable whether it should be a BadPaddingException. I would rather use an IllegalArgumentException there.